Citation dialog: warn about cross-library citations

abaevbog opened the pull request:
When a new item is added into the citation, and all other cited items are all from another library, display a warning saying that cross-library citations are not recommended. If the document has cited items from multiple libraries, no warning is shown.

If there are many cited items, and `io` is still loading, the warning will appear if needed as soon as `io` is ready.

If the user clicks "Cancel" in the warning dialog, all added items from other libraries are removed.

Fixes: #5631

---

For reference:

https://github.com/user-attachments/assets/0667ee52-0c7c-4ebc-ad70-7c79fb4408b5

If io takes a long time to load, and the user adds some items and clicks `Accept`, the warning will be shown before the dialog is accepted:

https://github.com/user-attachments/assets/f04800e5-9826-4c5c-9c49-e761ce714b68




I wasn't sure about the use case for the pref to not show this warning in this document again, as mentioned in the linked issue. Once the user adds at least one citation from another library, next time the dialog is used, all cited items will not belong to a single library, so the warning won't be shown already.

adomasven commented on chrome/content/zotero/integration/citationDialog.js:
```diff
@@ -1218,6 +1249,53 @@ const IOManager = {
+		let citedLibraries = new Set(citedItems.map(item => item.libraryID).filter(l => l));
+		if (citedLibraries.size !== 1) return true;
+		// If the items being added belong to the same library as all cited items, do nothing
+		let citedLibraryID = citedLibraries.values().next().value;
```
Won't this return the libraryID of the newly added citation if it's being cited as the first item of the first citation in the document, i.e. being citation insert before all other citations in the doc?

abaevbog commented:
Now that I have tested this more, though, I feel like the warning itself isn't particularly helpful...

Suppose you cited only from Group A in a document, and you want to cite an item, say, opened in the reader tab that belongs to Group B. You select it, get a warning that it's not recommended, and ... then what? Do you go back to the main window, drag the item from group B to group A, and then go back to the citation dialog to cite that copied item from group A? Feels like a lot of extra steps.

I wonder if we should also add an option in the warning dialog to copy the item to group A and cite that instead? Essentially, it would save the user the trip to the main Zotero window and make the process a bit smoother.

In that case, though, the warning could potentially appear multiple times for the same document, which is certainly more annoying. Could we also consider adding a document-level pref of a "Target Group"? If that is set, and one is about to cite an item from a different library, we could copy the item to the "Target Group" and cite that right away without showing the warning dialog. I think it might also help with some of the issues the citation explorer is meant to fix, since it should make it more likely that all items come from the same group (when it is applicable)?

@adomasven, @dstillman, what do you think about this idea?

abaevbog commented:
Along the same lines, when the user is adding an unlinked item from the "Cited" section, we could suggest to save the item to their library, or just save it right away if the pref is set. That way, it should reduce likelihood that one will end up with a bunch of unlinked items in citations.

dstillman commented:
To clarify, I was saying we would automatically maintain the list of used libraries, not that we would expose it as a distinct pref or let users choose it explicitly. The user's choice is implicit based on their first selection (and if they messed up, that's what the Citation Explorer is for).

abaevbog committed:
citation dlg: warn about cross-library citations

- new Zotero.Integration.refreshCitedLibraries function
to record the libraries that are currently cited in the document
- when a new item is about to be cited, show a warning
if it does not belong to currently cited libraries
- cited libraries are saved in the document preferences
so that they persist across sessions
- cited libraries are recorded as group URIs and their
names. That way, a user that does not have access to a cited
group can still receive a warning with the group's name.
Also, URIs allow us to distinguish between "My Library"-s
of different users.
- cited libraries in library mode appear at the top of the
collectionTree
- in list mode, cited libraries receive priority in search results,
so that they always appear after selected/open/cited items.

abaevbog committed:
test edits

yexingsha commented:
I think we can add a "Cited" label to cited libraries. This does not change the structure or order of the collection tree, and works for the list mode as well:

<img width="896" height="596" alt="library" src="https://github.com/user-attachments/assets/7ffa17bd-ee37-4b5b-833e-e40d8f3df9f5" />

<img width="896" height="546" alt="list" src="https://github.com/user-attachments/assets/6481c5e5-3b6b-4c61-a57d-f95c672cd926" />

In library mode, when the row is focused, use `#FFFFFF33` as the label's background color, and `accent-white` as its text color.

abaevbog commented:
> This does not change the structure or order of the collection tree, and works for the list mode as well:

I feel like changing the order of collection tree is a significant part of the idea here... In list mode, after selected/open/cited sections, the cited libraries go first before remaining not-cited libraries. Similarly, it feels appropriate to make cited libraries "special" by placing them at the top, which implies that these are the libraries you are expected to be citing from. Some people end up having a fairly large number of groups (e.g. several tens), in which case it's also more useful to have the currently-cited libraries at the top vs having to scroll through the entire list to locate them.

abaevbog committed:
set default io.getCitedLibraryInfo if not provided

Fix breakage when opening dialog from note editor

abaevbog committed:
set cited libraries after io is loaded

To account for many instances when cited libraries
from document data may be outdated or missing

yexingsha commented:
> Some people end up having a fairly large number of groups (e.g. several tens), in which case it's also more useful to have the currently-cited libraries at the top vs having to scroll through the entire list to locate them.

Can't we just sort the cited libraries to the top of the group libraries? I think I would prefer to keep the My Library / Group Libraries structure consistent, and also use the same visual indicator for both library mode and list mode. But that's mostly an aesthetic preference.

If we do put the cited libraries into a group at the top, we can use `16/universal/cite.svg` for the icon and color it `accent-blue`, and put 8px spacing between the last library and the divider.

Also, this just occurred to me: can we put a "Don't warn me again in this document" checkbox in the dialog, and an equivalent one in Document Preferences? So that when people don't write with a collaborator, they can just cite from wherever they want.

abaevbog committed:
keep collectionTree structure, add "Cited" label

- order of collectionTree groups does not change
- add "Cited" badge after cited groups in collectionTree
and in list mode
- when citation dialog opens, non-cited groups are
collapsed, even if they were expanded before last
close

abaevbog committed:
use sessionID from doc data to create new session

And, if session has a record of cited library URIs,
do not override it with document data, if it is empty.

It makes it more likely that one will be able to
maintain a record of cited libraries even if another
collaborator is on an old version of Zotero that will not
save cited libraries to document data

abaevbog committed:
minor collection tree cleanup

abaevbog committed:
fix blinking cited badge in collectionTree

yexingsha commented:
> This might be an argument in favor if putting the spinner after "Cited" badge in collectionTree. We can also display the "Cited" badge as faint-grey-ish until io is loaded, and change it to blue once we confirm that the group is indeed 100% cited.

Sounds good to me. It can look like this:

<img width="896" height="596" alt="loading" src="https://github.com/user-attachments/assets/61376b57-c407-41b3-9988-ef4328c2cc1c" />

- The spinner is put before the text so that it doesn't shift when the spinner disappears.
- In this state, the badge background is `fill-quarternary`, text is `fill-secondary`, and spinner is `fill-tertiary`.
- The blue that it changes to should be `accent-blue30`. Right now the blue is a bit too dark.

abaevbog committed:
add loading state to the cited badge

Loading state has darker colors, a spinner, and a tooltip
